File: Cores/DotNet/Random/Statistics.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;

namespace OSOL.Extremum.Cores.DotNet.Random
{
    public static class Statistics
    {

        public static double SampleMoment(IEnumerable<double> data, int n) =>
            data.Select(x => Math.Pow(x, n)).Average();

        public static double GetMean(IEnumerable<double> data) => SampleMoment(data, n: 1);

        public static double CentralMoment(IEnumerable<double> data, int n)
        {
            double[] dataArray = data as double[] ?? data.ToArray();
            double average = GetMean(dataArray);
            return dataArray.Select(x => Math.Pow(x - average, n)).Average();
        }

        public static double GetUnbiasedSigma(IEnumerable<double> data)
        {
            double[] dataArray = data as double[] ?? data.ToArray();
            double size = dataArray.Count();
            return Math.Sqrt(CentralMoment(dataArray, 2) * (size / (size - 1)));
        }

    }
}

File: Cores/JVM/src/main/scala/OSOL/Extremum/Cores/JVM/Random/Statistics.scala
package OSOL.Extremum.Cores.JVM.Random

object Statistics {

  def sampleMoment(data: Traversable[Double], n: Int): Double =
    data.map(x => math.pow(x, n)).sum / data.size

  def getMean(data: Traversable[Double]): Double = sampleMoment(data, 1)

  def centralMoment(data: Traversable[Double], n: Int): Double = {
    val average = getMean(data)
    data.map(x => math.pow(x - average, n)).sum / data.size
  }

  def getUnbiasedSigma(data: Traversable[Double]): Double = math.sqrt((data.size / (data.size - 1.0)) * centralMoment(data, 2))

}
